File: python/749.py
"""Project Euler Problem 749: Near Power Sums.

Find the sum of the near power sums (the sum of the kth powers of its digits
is 1 away from the number itself, for some k) with up to N digits.

We brute force over all combinations of digits, with the following two
optimizations:
- k must be even. Otherwise, Σ (d_i)^k ≡ Σ d_i ≡ Σ (10^i) d_i (mod 3), so
  they cannot be 1 apart.
- The total number of digits cannot be more than 1 greater than k, because
  then it is not possible for even a number with all 9s to add up to a large
  enough sum.
"""

from __future__ import annotations


def imod(a: int, m: int) -> int:
    """Modulo operation: a mod m."""
    return a % m


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Fast exponentiation modulo mod."""
    result = 1
    base = base % mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def nth_pows(base: int, k: int, max_val: int) -> list[int]:
    """Precompute kth powers of digits."""
    result = []
    for d in range(base):
        pow_val = d**k
        if pow_val > max_val:
            break
        result.append(pow_val)
    return result


def solve() -> int:
    """Solve Problem 740."""
    n = 16
    b = 10
    nums: set[int] = set()

    def helper(d: int, num_digits: int, sum_powers: int, k: int, pows: list[int]) -> None:
        """Recursive helper."""
        for candidate in [sum_powers - 1, sum_powers + 1]:
            if candidate > 0:
                actual_sum_powers = 0
                num = candidate
                while num > 0:
                    actual_sum_powers += pows[imod(num, b)]
                    num //= b
                if actual_sum_powers == sum_powers:
                    nums.add(candidate)

        if num_digits < n and num_digits <= k + 1:
            for new_d in range(d, b):
                new_sum = sum_powers + pows[new_d]
                if new_sum < b**n:
                    helper(new_d, num_digits + 1, new_sum, k, pows)

    for k in range(2, n + 2, 2):
        pows = nth_pows(b, k, b**n)
        helper(1, 0, 0, k, pows)

    return sum(nums)


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/749.c
/* Project Euler Problem 749: Near Power Sums.
 * Translated from python/749.py
 *
 * Find the sum of near power sums (sum of kth powers of digits is 1 away
 * from the number itself) with up to N digits.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long ll;
typedef unsigned long long ull;

#define MAXN 16
#define BASE 10
#define MAX_NUMS 100000

static ull nums[MAX_NUMS];
static int num_count = 0;

/* Precomputed kth powers of digits */
static ull pows[BASE];

static ull pow_limit; /* 10^N */

/* Check if candidate is a near power sum for the given k */
static int check_candidate(ull candidate, int k) {
    ull sum = 0;
    ull num = candidate;
    while (num > 0) {
        int d = num % BASE;
        sum += pows[d];
        num /= BASE;
    }
    return sum == candidate + 1 || sum == candidate - 1;
    (void)k;
}

static void helper(int d, int num_digits, ull sum_powers, int k) {
    /* Check sum_powers - 1 and sum_powers + 1 */
    for (int delta = -1; delta <= 1; delta += 2) {
        ull candidate = sum_powers + delta;
        if (candidate > 0 && candidate < pow_limit) {
            /* Verify: sum of kth powers of digits of candidate == sum_powers */
            ull actual_sum = 0;
            ull num = candidate;
            while (num > 0) {
                int dig = num % BASE;
                actual_sum += pows[dig];
                num /= BASE;
            }
            if (actual_sum == sum_powers) {
                /* Add to set (check for duplicates later) */
                if (num_count < MAX_NUMS) {
                    nums[num_count++] = candidate;
                }
            }
        }
    }

    if (num_digits < MAXN && num_digits <= k + 1) {
        for (int new_d = d; new_d < BASE; new_d++) {
            ull new_sum = sum_powers + pows[new_d];
            if (new_sum < pow_limit) {
                helper(new_d, num_digits + 1, new_sum, k);
            }
        }
    }
}

int cmp_ull(const void *a, const void *b) {
    ull va = *(const ull *)a;
    ull vb = *(const ull *)b;
    if (va < vb) return -1;
    if (va > vb) return 1;
    return 0;
}

int main() {
    pow_limit = 1;
    for (int i = 0; i < MAXN; i++) pow_limit *= BASE;

    for (int k = 2; k <= MAXN + 2; k += 2) {
        /* Compute kth powers of digits */
        for (int d = 0; d < BASE; d++) {
            ull p = 1;
            for (int j = 0; j < k; j++) {
                p *= d;
                if (p >= pow_limit) { p = pow_limit; break; }
            }
            pows[d] = p;
        }
        helper(1, 0, 0, k);
    }

    /* Sort and deduplicate */
    qsort(nums, num_count, sizeof(ull), cmp_ull);

    ull total = 0;
    for (int i = 0; i < num_count; i++) {
        if (i == 0 || nums[i] != nums[i - 1]) {
            total += nums[i];
        }
    }

    printf("%llu\n", total);
    return 0;
}
